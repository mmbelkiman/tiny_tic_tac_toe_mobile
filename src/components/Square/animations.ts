import {Animated} from 'react-native';

export const playShakeAnimation = (animation: Animated.Value) => {
  Animated.sequence([
    Animated.timing(animation, {
      toValue: 1,
      duration: 100,
      useNativeDriver: true,
    }),
    Animated.timing(animation, {
      toValue: 0,
      duration: 100,
      useNativeDriver: true,
    }),
  ]).start();
};

export const getShakeStyle = (animation: Animated.Value) => ({
  transform: [
    {
      translateX: animation.interpolate({
        inputRange: [0, 0.2, 0.4, 0.6, 0.8, 1],
        outputRange: [0, -5, 5, -5, 5, 0],
      }),
    },
    {
      translateY: animation.interpolate({
        inputRange: [0, 0.2, 0.4, 0.6, 0.8, 1],
        outputRange: [-2, 0, 2, 0, -2, 0],
      }),
    },
  ],
});
