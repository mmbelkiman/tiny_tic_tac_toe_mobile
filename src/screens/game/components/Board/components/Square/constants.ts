export const BORDER_COLOR = '#fff';
export const BORDER_WIDTH = 2;
export const BACKGROUND_COLOR = '#00000011';
