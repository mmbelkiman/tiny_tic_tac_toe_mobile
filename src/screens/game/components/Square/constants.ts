export const BORDER_COLOR = '#00000022';
export const BACKGROUND_COLOR = '#33333311';
