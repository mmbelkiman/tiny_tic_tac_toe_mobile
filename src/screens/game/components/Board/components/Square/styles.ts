import {StyleSheet} from 'react-native';
import {BACKGROUND_COLOR, BORDER_COLOR, BORDER_WIDTH} from './constants';

export default StyleSheet.create({
  square: {
    borderWidth: BORDER_WIDTH,
    borderColor: BORDER_COLOR,
    backgroundColor: BACKGROUND_COLOR,
  },
});
