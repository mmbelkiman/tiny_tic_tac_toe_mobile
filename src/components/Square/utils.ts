import styles from './styles';
import {getShakeStyle} from './animations';
import {Animated} from 'react-native';

export const getSquareStyle = (
  squareSize: number,
  animation: Animated.Value,
) => [
  styles.square,
  {width: squareSize, height: squareSize},
  getShakeStyle(animation),
];
