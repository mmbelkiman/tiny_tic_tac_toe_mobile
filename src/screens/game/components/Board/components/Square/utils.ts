import {Animated} from 'react-native';
import styles from './styles';

export const getShakeStyle = (animation: Animated.Value) => ({
  transform: [
    {
      translateX: animation.interpolate({
        inputRange: [0, 0.2, 0.4, 0.6, 0.8, 1],
        outputRange: [0, -5, 5, -5, 5, 0],
      }),
    },
    {
      translateY: animation.interpolate({
        inputRange: [0, 0.2, 0.4, 0.6, 0.8, 1],
        outputRange: [-2, 0, 2, 0, -2, 0],
      }),
    },
  ],
});

export const getSquareStyle = (
  squareSize: number,
  animation: Animated.Value,
) => [
  styles.square,
  {width: squareSize, height: squareSize},
  getShakeStyle(animation),
];
