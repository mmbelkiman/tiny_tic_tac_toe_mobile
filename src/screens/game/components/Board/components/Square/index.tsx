import React, {useRef, useState} from 'react';
import {Animated, TouchableOpacity} from 'react-native';
import {PlayerMark} from '@game/common/types';
import {playShakeAnimation} from './animations';
import LottiePlayerMark from './components/lottiePlayerMark';
import {getPlayerMarkColor, getSquareStyle} from './utils';

interface SquareProps {
  playerMarkIsCircle: boolean;
  onPress: () => void;
  squareSize: number;
  disabled: boolean;
  circleColor: string;
  crossColor: string;
  testID?: string;
}

const Square: React.FC<SquareProps> = ({
  playerMarkIsCircle,
  squareSize,
  disabled,
  onPress,
  circleColor,
  crossColor,
  testID = '',
}) => {
  const animation = useRef(new Animated.Value(0)).current;
  const [hasPlayedAnimations, setHasPlayedAnimations] = useState(false);

  //It is necessary to persist the Player Mark state to prevent it from changing when a new prop comes from the board.
  const [isCircle, setIsCircle] = useState<boolean>(false);

  const handleOnPress = () => {
    playShakeAnimation(animation);

    //Skip other effects and actions if already has player set
    if (hasPlayedAnimations) {
      return;
    }
    setHasPlayedAnimations(true);
    setIsCircle(playerMarkIsCircle);
    onPress();
  };

  return (
    <TouchableOpacity
      testID={`Square${testID}.TouchableOpacity`}
      disabled={disabled}
      onPress={handleOnPress}>
      <Animated.View style={getSquareStyle(squareSize, animation)}>
        <LottiePlayerMark
          color={getPlayerMarkColor({
            isCircle,
            crossColor,
            circleColor,
          })}
          isCircle={isCircle}
          isVisible={hasPlayedAnimations}
          squareSize={squareSize}
        />
      </Animated.View>
    </TouchableOpacity>
  );
};

export default Square;
