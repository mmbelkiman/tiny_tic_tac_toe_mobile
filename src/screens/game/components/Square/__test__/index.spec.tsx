import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import Square from '../index';

describe('Square Component', () => {
  it('should call the onPress just at first press', () => {
    const onPressMock = jest.fn();
    const {getByTestId} = render(
      <Square
        playerMark="X"
        squareSize={100}
        winner={null}
        onPress={onPressMock}
      />,
    );

    const square = getByTestId('Square.TouchableOpacity');
    fireEvent.press(square);
    fireEvent.press(square);
    fireEvent.press(square);

    expect(onPressMock).toBeCalledTimes(1);
  });

  it('should not allow pressing when there is a winner', () => {
    const onPressMock = jest.fn();
    const {getByTestId} = render(
      <Square
        playerMark="X"
        squareSize={100}
        winner="X"
        onPress={onPressMock}
      />,
    );

    const square = getByTestId('Square.TouchableOpacity');
    fireEvent.press(square);

    expect(onPressMock).not.toHaveBeenCalled();
  });

  it('renders correctly with playerMark "X" and squareSize 100', () => {
    const {toJSON} = render(
      <Square
        playerMark="X"
        squareSize={100}
        winner={null}
        onPress={() => {}}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with playerMark "O" and squareSize 32', () => {
    const {toJSON} = render(
      <Square
        playerMark="O"
        squareSize={32}
        winner={null}
        onPress={() => {}}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with a winner', () => {
    const {toJSON} = render(
      <Square playerMark="X" squareSize={100} winner="X" onPress={() => {}} />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly without winner and playerMark', () => {
    const {toJSON} = render(
      <Square
        playerMark=""
        squareSize={100}
        winner={null}
        onPress={() => {}}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
