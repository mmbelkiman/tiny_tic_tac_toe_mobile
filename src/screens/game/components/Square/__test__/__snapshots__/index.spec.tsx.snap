// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Square Component renders correctly with a winner 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
  testID="Square.TouchableOpacity"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#33333311",
        "borderColor": "#00000022",
        "borderWidth": 1,
        "height": 100,
        "transform": [
          {
            "translateX": 0,
          },
          {
            "translateY": -2,
          },
        ],
        "width": 100,
      }
    }
  />
</View>
`;

exports[`Square Component renders correctly with playerMark "O" and squareSize 32 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
  testID="Square.TouchableOpacity"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#33333311",
        "borderColor": "#00000022",
        "borderWidth": 1,
        "height": 32,
        "transform": [
          {
            "translateX": 0,
          },
          {
            "translateY": -2,
          },
        ],
        "width": 32,
      }
    }
  />
</View>
`;

exports[`Square Component renders correctly with playerMark "X" and squareSize 100 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
  testID="Square.TouchableOpacity"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#33333311",
        "borderColor": "#00000022",
        "borderWidth": 1,
        "height": 100,
        "transform": [
          {
            "translateX": 0,
          },
          {
            "translateY": -2,
          },
        ],
        "width": 100,
      }
    }
  />
</View>
`;

exports[`Square Component renders correctly without winner and playerMark 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
  testID="Square.TouchableOpacity"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#33333311",
        "borderColor": "#00000022",
        "borderWidth": 1,
        "height": 100,
        "transform": [
          {
            "translateX": 0,
          },
          {
            "translateY": -2,
          },
        ],
        "width": 100,
      }
    }
  />
</View>
`;
